provisioner/terraform/executor.go

```diff
@@ -0,0 +1,338 @@
+package terraform
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"golang.org/x/xerrors"
+
+	"github.com/hashicorp/go-version"
+	tfjson "github.com/hashicorp/terraform-json"
+
+	"github.com/coder/coder/provisionersdk"
+	"github.com/coder/coder/provisionersdk/proto"
+)
+
+type executor struct {
+	binaryPath string
+	cachePath  string
+	workdir    string
+}
+
+func (e executor) basicEnv() []string {
+	// Required for "terraform init" to find "git" to
+	// clone Terraform modules.
+	env := os.Environ()
+	// Only Linux reliably works with the Terraform plugin
+	// cache directory. It's unknown why this is.
+	if e.cachePath != "" && runtime.GOOS == "linux" {
+		env = append(env, "TF_PLUGIN_CACHE_DIR="+e.cachePath)
+	}
+	return env
+}
```

Comment:
We seem to use this in every call. Instead of piping it through the execution functions, we could have a single `command` function that returns a command initialized with the `binaryPath`, `args`, and this env already applied.

Reply:
`plan()` and `apply()` have their own environment that include parameters from end users and the coder terraform provider stuff, so not every call has the same environment in the code I'm refactoring.  I _could_ unify things, but that ends up having to either initialize the `executor` with the parameters, or pass the parameters instead of `env` on every call.  Right now the `executor` only asks for the environment on `plan()` and `apply()` and uses this basic env for every other call.

Reply:
Makes sense!